File: apps/cargo-tracker/packs/handling/app/domain/handling/domain/cargo_snapshot.rb
# frozen_string_literal: true

module Handling
  module Domain
    # 貨物スナップショット（ACL・US15）。Booking Context の公開ビューから射影した読み取り専用の値。
    # 荷役妥当性（route_check）の照合に用いる。Handling は Booking 集約に直接依存しない（ADR-0001/0003）。
    CargoSnapshot = Data.define(:booking_id, :origin, :destination, :leg_locations) do
      def initialize(booking_id:, origin:, destination:, leg_locations: [])
        super(booking_id: booking_id, origin: origin, destination: destination,
              leg_locations: Array(leg_locations))
      end

      # 旅程が経由する港（積地・揚地）の集合。旅程未紐付けなら出発地・目的地のみ。
      def route_ports
        (leg_locations + [ origin, destination ]).compact.uniq
      end
    end
  end
end

File: docs/article/source/java-4/apps/cargo-tracker/backend/handlingms/src/main/java/com/example/cargotracker/handling/domain/model/valueobjects/CargoSnapshot.java
package com.example.cargotracker.handling.domain.model.valueobjects;

import com.example.cargotracker.shared.domain.location.Location;
import java.util.List;

/**
 * 貨物の写し（ACL の読み取りモデル / domain-model.md「Handling Context」）。
 *
 * <p><b>予定ルート外かどうかは、ここが 1 か所で答える。</b> 集約も画面もこの判定を
 * 書き直さない。書き直すと判定が 2 つになり、片方だけ直る。</p>
 *
 * <p><b>Booking / Tracking の型を持ち込まない。</b> 港は共有カーネルの
 * {@link Location} で表す。</p>
 */
public record CargoSnapshot(
        String trackingNumber,
        String bookingId,
        Location origin,
        Location destination,
        String cargoType,
        List<LegSnapshot> legs) {

    public CargoSnapshot {
        legs = legs == null ? List.of() : List.copyOf(legs);
    }

    /** 予定の旅程の 1 区間。<b>時刻は持たない</b>（[ADR-0012] 決定 4）。 */
    public record LegSnapshot(String voyageNumber, Location load, Location unload) {
    }

    /**
     * その種別をその場所で行うのは予定外か（不変条件 2・3）。
     *
     * <p>照合する港は種別が決める（domain-model.md「荷役種別ごとの要件」）。</p>
     *
     * <ul>
     *   <li>{@code RECEIVE} … 出発港</li>
     *   <li>{@code LOAD} … 旅程のどこかの積込港</li>
     *   <li>{@code UNLOAD} … 旅程のどこかの荷降港</li>
     *   <li>{@code CLAIM} … 目的港</li>
     * </ul>
     *
     * <p><b>旅程が無いなら予定外に倒す</b>（不変条件 3）。分からないときは
     * 「予定どおり」と言わない——予定外の記録は残るが、予定どおりと記録すると
     * 誤配が見えなくなる。</p>
     */
    public boolean isOffRoute(HandlingType type, Location location) {
        if (location == null) {
            return true;
        }
        return switch (type) {
            case RECEIVE -> !location.equals(origin);
            case CLAIM -> !location.equals(destination);
            case LOAD -> legs.isEmpty()
                    || legs.stream().noneMatch(leg -> location.equals(leg.load()));
            case UNLOAD -> legs.isEmpty()
                    || legs.stream().noneMatch(leg -> location.equals(leg.unload()));
        };
    }
}
